core/src/parseURL.ts

```diff
@@ -10,43 +12,76 @@ export class ParseURLError extends Error {
-    /** `label` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#label) */
-    label: string | undefined;
-    /** `message` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#message) */
-    message: string | undefined;
-    /** `memo` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#memo) */
-    memo: string | undefined;
+export interface TransactionRequestURL {
+    /** `link` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#link). */
+    link: Link;
+    /** `label` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#label). */
+    label: Label | undefined;
+    /** `message` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#message). */
+    message: Message | undefined;
 }
 
 /**
- * Parse the components of a Solana Pay URL.
+ * A Solana Pay transfer request URL.
+ */
+export interface TransferRequestURL {
+    /** `recipient` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#recipient). */
+    recipient: Recipient;
+    /** `amount` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#amount). */
+    amount: Amount | undefined;
+    /** `spl-token` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#spl-token). */
+    splToken: SPLToken | undefined;
+    /** `reference` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#reference). */
+    reference: Reference[] | undefined;
+    /** `label` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#label). */
+    label: Label | undefined;
+    /** `message` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#message). */
+    message: Message | undefined;
+    /** `memo` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#memo). */
+    memo: Memo | undefined;
+}
+
+/**
+ * Parse a Solana Pay URL.
  *
- * **Reference** implementation for wallet providers.
+ * @param url - URL to parse.
  *
- * @param url - A Solana Pay URL
+ * @throws {ParseURLError}
  */
-export function parseURL(url: string): ParsedURL {
-    if (url.length > 2048) throw new ParseURLError('length invalid');
+export function parseURL(url: string | URL): TransactionRequestURL | TransferRequestURL {
+    if (typeof url === 'string') {
+        if (url.length > 2048) throw new ParseURLError('length invalid');
+        url = new URL(url);
+    }
+
+    const { protocol, pathname } = url;
+    if (protocol !== SOLANA_PROTOCOL) throw new ParseURLError('protocol invalid');
+    if (!pathname) throw new ParseURLError('pathname missing');
+
+    return /[^A-Za-z0-9]/.test(pathname) ? parseTransactionRequestURL(url) : parseTransferRequestURL(url);
```

Comment:
Question: does this work because `pathname` for a transaction request is an account, where for a transfer request it's a URL and must have some non-alphanumeric characters?

Reply:
Refactored to `/[:%]/` since it's simpler, and if `.sol` or something becomes supported in the future, this won't interpret it as a transaction request URL.